File: 209.长度最小的子数组.java
/*
 * @lc app=leetcode.cn id=209 lang=java
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
class Solution {
    // slide window
    // when the left update and the right stop
    public int minSubArrayLen(int target, int[] nums) {
        int l = 0, r = 0;
        int minLength = 1000000;
        int sum = 0;

        for(; r < nums.length; r++) {
            // udpate the state
            sum += nums[r];
            while(sum >= target) { // left will go move condition
                int length = r - l  + 1; // current length 
                minLength = minLength < length ? minLength : length;
                sum -= nums[l++];
            }

        }
        if(minLength == 1000000) return 0;
        return minLength;
    }
}
// @lc code=end

File: 209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        // 滑动窗口？
        int left = 0;
        int sum = 0, res = INT_MAX;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            while(sum >= target) {
                int length = i - left + 1;
                res = res > length ? length : res;
                sum -= nums[left];
                left++;
            } 
        }
        return res == INT_MAX ? 0 : res;
    }
};
// @lc code=end

